### ISSUE TITLE
rpc help helper class

### ISSUE THREAD

**Issue by MarcoFalke** [bitcoin/bitcoin#14378]
The rpc help (documentation) is currently manually updated with no checks of correctness or consistency with the code.

I think the goal should be that the rpc help is generated by pretty printing a cpp object (as opposed to the hardcoded string)

In a first step, the constructor of the class could just take the name of the method, and an array of args. Each arg would be a json value (c.f. https://gist.github.com/karel-3d/1490786786525b0365ea8f459a9fc683#examples-1) with the type and the name to be used when calling with named arguments. Also, each arg would have flags for optional/required (and a default for non-required args), as well as a short description string.

Further enhancements would be to do the same for the result (returned json) and finally to check the types provided in the documentation against the types passed during runtime.

**Comment by karel-3d** [comment]
I have only now found out this issue and your PR

I have made a separate issue here

https://github.com/bitcoin/bitcoin/issues/14399

If you make a helper class, the results of the parser can probably be directly used (with some changes - removing the markdown=>html converter for example).

**Comment by MarcoFalke** [comment]
Indeed, you could just teach the helper class to export in md or html or whatever instead of our internal format.

**Comment by achow101** [comment]
I've been working on a helper class for this but has turned out to be more complicated than I expected. I hope to get a pr for it out in a few days.

**Comment by karel-3d** [comment]
If it helps, I made slight change to the parser and exported the RPC to JSON, with the structure logic of the proposal

https://gist.github.com/karel-3d/02f8143b8fd463f8fd2b74a9044614c1

**Comment by etscrivner** [comment]
Put together a small prototype for this using a builder interface [here](https://gist.github.com/etscrivner/ffac70de141c7b29441a3156fe1271d6). A small example is below:

```c++
auto docs = RPCDocumentation("getrawmempool")
  .DescriptionLine("Returns all transaction ids in memory pool as a json array of string transaction ids.")
  .DescriptionLine("Hint: use getmempoolentry to fetch a specific transaction from the mempool.")
  .Argument("first", "integer", "The first argument is required", true)
  .Argument("verbose", "boolean", "True for json object, false for array of transaction ids", false, "false")
  .Argument("other", "string", "Another argument");
```

You can then generate docs as follows:

```c++
std::cout << docs.write()
```

The result looks as follows:

```
getrawmempool first ( verbose "other" )
Returns all transaction ids in memory pool as a json array of string transaction ids.

Hint: use getmempoolentry to fetch a specific transaction from the mempool.

Arguments:
1. first (integer) The first argument is required
2. verbose (boolean, optional, default=false) True for json object, false for array of transaction ids
3. other (string, optional) Another argument
```

While this is only a small first step, curious if this is what @MarcoFalke had in mind or something different?

**Comment by karel-3d** [comment]
Just a node - some of the RPCs have very wild docs in the input arguments.

See `fundrawtransaction` for the big doc with respect to the input.

https://bitcoin-rpc.github.io/en/doc/0.17.99/rpc/rawtransactions/fundrawtransaction/ (made with my other PR)

**Comment by achow101** [comment]
Ideally the helper class should be able to work for handling the help for JSON formatted arguments and results. IMO such a helper is most useful for formatting the results as those are usually the things that have the most indentation and whitespace issues as they are typically large JSON things.

I've begun working on this here: https://github.com/achow101/bitcoin/tree/rpc-help-univ.

**Comment by karel-3d** [comment]
@etscrivner I got inspired by your gist and I edited a little to match what I actually found in the RPC docs and how the RPC docs look like

https://gist.github.com/karel-3d/5847ea0172350368dead323211893faa

I will try to actually add that to bitcoin source code soon-ish and make a PR.

(By looking in the source code, I found out I missed a lot of "hidden" RPCs in all the extraction scripts. I didn't even know they exist!)

**Comment by domob1812** [comment]
FYI, Namecoin already uses a [class with a fluent interface](https://github.com/namecoin/namecoin-core/blob/db0bcba13273d5b83d80cb8077861ae8744f80bb/src/rpc/names.h#L58) (similar to what @etscrivner suggested above) to build some help texts, altough in a particularly repetetive setting.

**Comment by MarcoFalke** [comment]
Looks like that doesn't work with types. I.e. you'd still have to manually stringify the type name and adjust the formatting based on the type.

https://github.com/namecoin/namecoin-core/blob/db0bcba13273d5b83d80cb8077861ae8744f80bb/src/rpc/names.cpp#L378-L380

Also, it doesn't work with pretty printing dictionaries or arrays:

https://github.com/namecoin/namecoin-core/blob/db0bcba13273d5b83d80cb8077861ae8744f80bb/src/rpc/names.cpp#L357-L359

**Comment by domob1812** [comment]
Well, that's just what the code is at the moment (in case that wasn't clear, I wrote it).  The main goal for it was to just simplify some part of the formatting that was particularly repetetive.  The interface can surely be extended (and adapted for Bitcoin) to pretty print dictionaries and other things more nicely.

**Comment by karel-3d** [comment]
great work

**Comment by jnewbery** [comment]
This was closed by #14796
